### PR TITLE
Fine-granular enabling/disabling of plan-level optimizations

### PR DESCRIPTION
**Pull request by rschu1ze**
See the individual commits for what changed.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Plan-level optimizations can now be enabled/disabled individually. Previously, it was only possible to disable them all. The setting which previously did that (`query_plan_enable_optimizations`) is retained and can still be used to disable all optimizations.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 0a9f6467395c740a59ede9582a18c46f4e32881f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57152/0a9f6467395c740a59ede9582a18c46f4e32881f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6971890504">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57152/0a9f6467395c740a59ede9582a18c46f4e32881f/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -4805,6 +4805,183 @@ a	Tuple(
 
 If a table has a space-filling curve in its index, e.g. `ORDER BY mortonEncode(x, y)`, and the query has conditions on its arguments, e.g. `x >= 10 AND x <= 20 AND y >= 20 AND y <= 30`, use the space-filling curve for index analysis.
 
+## query_plan_enable_optimizations {#query_plan_enable_optimizations}
```
Comment (devcrafter) [inline docs/en/operations/settings/settings.md:4808]: Not sure about documenting the settings. They are too low level for users to deal with. So, having them is good for debugging and disabling some broken optimizations, but for this having them in settings is just enough.

cc @KochetovNicolai

**Review by devcrafter — APPROVED** [review]
LGTM with comment

File: docs/en/operations/settings/settings.md (hunk 1 above)
Comment (rschu1ze) [reply docs/en/operations/settings/settings.md:4808]: I generally like to read the markdown/html docs about settings. It is hard to draw a hard line between public and dev settings, unfortunately.

In particular setting `query_plan_enable_optimizations` (which existed before) was undocumented. One could argue that it is general-purpose enough to have it documented.

But I am so fine with not documenting the individual settings.

Comment (KochetovNicolai) [reply docs/en/operations/settings/settings.md:4808]: I am ok do document those settings. Maybe with some disclaimer that they are for debugging only, supposed to be used mainly by developers, do not imply any backward compatibility and might be changed/removed in any time.

Comment (rschu1ze) [reply docs/en/operations/settings/settings.md:4808]: Done, tnx.

**Pull request merged**